features/netsocket/cellular/generic_modem_driver/TESTS/unit_tests/default/main.cpp

```diff
@@ -0,0 +1,449 @@
+static void do_udp_echo(UDPSocket *sock, SocketAddress *host_address, int size)
+{
+    bool success = false;
+    void * recv_data = malloc(size);
+    TEST_ASSERT(recv_data != NULL);
+
+    // Retry this a few times, don't want to fail due to a flaky link
+    for (int x = 0; !success && (x < NUM_UDP_RETRIES); x++) {
+        tr_debug("Echo testing UDP packet size %d byte(s), try %d.", size, x + 1);
+        if ((sock->sendto(*host_address, (void*) test_data, size) == size)
+               && (sock->recvfrom(host_address, recv_data, size) == size)) {
+            TEST_ASSERT(memcmp(test_data, recv_data, size) == 0);
+            success = true;
+        }
+    }
+    TEST_ASSERT(success);
+
+    free(recv_data);
+}
+
+/**
+ * Send an entire TCP data buffer until done
+ */
+static int send_all(TCPSocket *sock, const char *data, int size)
+{
+    int x;
+    int count = 0;
+    Timer timer;
+
+    timer.start();
+    while ((count < size) && (timer.read_ms() < ASYNC_TEST_WAIT_TIME)) {
+        x = sock->send(data + count, size - count);
+        if (x > 0) {
+            count += x;
+            tr_debug("%d byte(s) sent, %d left to send.", count, size - count);
+        }
+        wait_ms(10);
+    }
+    timer.stop();
+
+    return count;
+}
+
+/**
+ * The asynchronous callback
+ */
+static void async_cb(bool *callback_triggered)
+{
+
+    TEST_ASSERT(callback_triggered != NULL);
+    *callback_triggered = true;
+}
+
+/**
+ * Do a TCP echo using the asynchronous driver
+ */
+static void do_tcp_echo_async(TCPSocket *sock, int size,
+                              bool *callback_triggered)
+{
+    void * recv_data = malloc(size);
+    int recv_size = 0;
+    int remaining_size;
+    int x, y;
+    Timer timer;
+    TEST_ASSERT(recv_data != NULL);
+
+    *callback_triggered = false;
+    tr_debug("Echo testing TCP packet size %d byte(s) async.", size);
+    TEST_ASSERT(send_all(sock, test_data, size) == size);
+    // Wait for all the echoed data to arrive
+    timer.start();
+    remaining_size = size;
+    while ((recv_size < size) && (timer.read_ms() < ASYNC_TEST_WAIT_TIME)) {
+        if (*callback_triggered) {
+            *callback_triggered = false;
+            x = sock->recv((char *) recv_data + recv_size, remaining_size);
+            if (x > 0) {
+                recv_size += x;
+                remaining_size = size - recv_size;
+                tr_debug("%d byte(s) echoed back so far, %d to go.", recv_size,
+                         remaining_size);
+            }
+        }
+        wait_ms(10);
+    }
+    TEST_ASSERT(recv_size == size);
+    y = memcmp(test_data, recv_data, size);
+    if (y != 0) {
+        tr_debug("Sent %d, |%*.*s|", size, size, size, test_data);
+        tr_debug("Rcvd %d, |%*.*s|", size, size, size, (char * ) recv_data);
+        // We do not assert a failure here because ublox TCP echo server doesn't send
+        // back original data. It actually constructs a ublox message string. They need to fix it as
```

Comment:
We use ublox's echo server for testing. It's a public echo server. However, unfortunately their TCP server doesn't echo back the data we send.  It sends a packet of the correct length but with arbitrary data. UDP server works as it should. This was reported to them already.